sql/core/src/main/scala/org/apache/spark/sql/Arrow.scala

```diff
@@ -195,13 +112,135 @@ object Arrow {
     val name = sparkField.name
     val dataType = sparkField.dataType
     val nullable = sparkField.nullable
+    new Field(name, nullable, sparkTypeToArrowType(dataType), List.empty[Field].asJava)
+  }
+}
 
-    dataType match {
-      case StructType(fields) =>
-        val childrenFields = fields.map(sparkFieldToArrowField).toList.asJava
-        new Field(name, nullable, ArrowType.Struct.INSTANCE, childrenFields)
-      case _ =>
-        new Field(name, nullable, getTypeFuncs(dataType).getType(), List.empty[Field].asJava)
+trait ColumnWriter {
+  def init(initialSize: Int): Unit
+  def write(data: Any): Unit
+  def finish(): (Seq[ArrowFieldNode], Seq[ArrowBuf])
+}
+
+/**
+ * Base class for flat arrow column writer, i.e., column without children.
+ */
+abstract class PrimitiveColumnWriter(protected val allocator: BaseAllocator) extends ColumnWriter {
+  protected val valueVector: BaseDataValueVector
+  protected val valueMutator: BaseMutator
+
+  var count = 0
+  var nullCount = 0
+
+  protected def writeNull()
+  protected def writeData(data: Any)
+  protected def valueBuffers(): Seq[ArrowBuf]
+  = valueVector.getBuffers(true) // TODO: check the flag
+
+  override def init(initialSize: Int): Unit = {
+    valueVector.allocateNew()
+  }
+
+  override def write(data: Any): Unit = {
+    if (data == null) {
+      writeNull()
+      nullCount += 1
+    } else {
+      writeData(data)
     }
+    count += 1
+  }
+
+  override def finish(): (Seq[ArrowFieldNode], Seq[ArrowBuf]) = {
+    valueMutator.setValueCount(count)
+    val fieldNode = new ArrowFieldNode(count, nullCount)
+    (List(fieldNode), valueBuffers)
+  }
+}
+
+class BooleanColumnWriter(allocator: BaseAllocator) extends PrimitiveColumnWriter(allocator) {
+  implicit def bool2int(b: Boolean): Int = if (b) 1 else 0
```

Comment:
There isn't a scala function to turn a boolean into a int, that's what is this for. I will change it to be non implicit